fuzz/ml-dsa.c

```diff
@@ -0,0 +1,624 @@
+
+    /*
+     * If `only_valid` is set, select only ML-DSA-44, ML-DSA-65, and ML-DSA-87.
+     * Otherwise, include some invalid sizes to trigger error paths.
+     */
+
+    if (only_valid)
+        modulus = 3;
+
+    /*
+     * Note, keylens for valid values (cases 0-2) are taken based on input
+     * values from our unit tests
+     */
+    switch (keysize % modulus) {
+    case 0:
+        *keytype = "ML-DSA-44";
+        *keylen = ML_DSA_44_PRIV_LEN;
+        break;
+    case 1:
+        *keytype = "ML-DSA-65";
+        *keylen = ML_DSA_65_PRIV_LEN;
+        break;
+    case 2:
+        *keytype = "ML-DSA-87";
+        *keylen = ML_DSA_87_PRIV_LEN;
+        break;
+    case 3:
+        /* select invalid alg */
+        *keytype = "ML-DSA-33";
+        *keylen = 33;
+        break;
+    case 4:
+        /* Select valid alg, but bogus size */
+        *keytype = "ML-DSA-87";
+        *buf = (uint8_t *)OPENSSL_load_u16_le(&keysize, *buf);
+        *len -= sizeof(uint16_t);
+        *keylen = (size_t)keysize;
+        *keylen %= ML_DSA_87_PRIV_LEN; /* size to our key buffer */
+        break;
+    default:
+        *keytype = NULL;
+        *keylen = 0;
+        break;
+    }
+    return 1;
+}
+
+/**
+ * @brief Creates an ML-DSA raw key from a buffer.
+ *
+ * This function selects a key type and size from the buffer, generates a random
+ * key of the appropriate length, and creates either a public or private ML-DSA
+ * key using OpenSSL's EVP_PKEY interface.
+ *
+ * @param buf   Pointer to the buffer pointer; updated after reading.
+ * @param len   Pointer to the remaining buffer size; updated accordingly.
+ * @param key1  Pointer to store the generated EVP_PKEY key (public or private).
+ * @param key2  Unused parameter (reserved for future use).
+ *
+ * @note The generated key is allocated using OpenSSL's EVP_PKEY functions
+ *       and should be freed appropriately using `EVP_PKEY_free()`.
+ */
+static void create_ml_dsa_raw_key(uint8_t **buf, size_t *len,
+                                  void **key1, void **key2)
+{
+    EVP_PKEY *pubkey;
+    char *keytype = NULL;
+    size_t keylen = 0;
+    /* MAX_ML_DSA_PRIV_LEN is longer of that and ML_DSA_87_PUB_LEN */
+    uint8_t key[MAX_ML_DSA_PRIV_LEN];
+    int pub = 0;
+
+    if (!select_keytype_and_size(buf, len, &keytype, &keylen, 0))
+        return;
+
+    /*
+     * Select public or private key creation based on the low order bit of the
+     * next buffer value.
+     * Note that keylen as returned from select_keytype_and_size is a public key
+     * length, so make the adjustment to private key lengths here.
+     */
+    if ((*buf)[0] & 0x1) {
+        pub = 1;
+    } else {
+        switch (keylen) {
+        case (ML_DSA_44_PUB_LEN):
+            keylen = ML_DSA_44_PRIV_LEN;
+            break;
+        case (ML_DSA_65_PUB_LEN):
+            keylen = ML_DSA_65_PRIV_LEN;
+            break;
+        case (ML_DSA_87_PUB_LEN):
+            keylen = ML_DSA_87_PRIV_LEN;
```

Comment:
These expect a public key length to convert to a private key length, but it should be other way around, with the switch in the `pub = 1` branch.

Reply:
After fixing the problem with `select_keytype_and_size`, then this existing code should be correct